### PR TITLE
Fix `GenericRateLimiter` hanging bug

### LINKED ISSUE #11742: Ratelimiter continues to Wait during Flush(or Compaction) and cannot be Signaled
### Question

I was doing a write with rocksdb and suddenly stopped writing. I looked at the log and found that it was because of too many memtables:

My program was using rocksdb(v8.0.0) for writing when it suddenly stopped writing. I looked at the LOG and found that it was because there were too many memtables:

```
2023/08/23-14:35:36.099044 139711273236224 [WARN] [db/column_family.cc:938] [metadata] Stopping writes because we have 4 immutable memtables (waiting for flush), max_write_buffer_number is set to 4
```

Then I printed out the stack and it is found that **rocksdb is blocked during Flush because `RateLimiter` continues to wait and cannot be signaled**. 

I took a closer look at the ratelimiter code. Still not sure what happens when all rate limiter request call `Wait` instead of `TimeWait` (and rocksdb guarantees that a request will always be signaled).

This problem can be recovered by restarting rocksdb.

### Stack

Here is the core call stack information:

```
TID 1675:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x000000000084940d rocksdb::port::CondVar::Wait()
#2  0x000000000090e321 rocksdb::GenericRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)
#3  0x000000000090d72d rocksdb::RateLimiter::RequestToken(unsigned long, unsigned long, rocksdb::Env::IOPriority, rocksdb::Statistics*, rocksdb::RateLimiter::OpType)
#4  0x00000000007f08af rocksdb::WritableFileWriter::WriteBuffered(char const*, unsigned long, rocksdb::Env::IOPriority)
#5  0x00000000007f3659 rocksdb::WritableFileWriter::Flush(rocksdb::Env::IOPriority)
#6  0x00000000007f4a0d rocksdb::WritableFileWriter::Append(rocksdb::Slice const&, unsigned int, rocksdb::Env::IOPriority)
#7  0x000000000087f0af rocksdb::BlockBasedTableBuilder::WriteMaybeCompressedBlock(rocksdb::Slice const&, rocksdb::CompressionType, rocksdb::BlockHandle*, rocksdb::BlockType, rocksdb::Slice const*)
#8  0x000000000088127e rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*, rocksdb::BlockType)
#9  0x0000000000881706 rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::BlockBuilder*, rocksdb::BlockHandle*, rocksdb::BlockType)
#10 0x0000000000882ad2 rocksdb::BlockBasedTableBuilder::Flush()
#11 0x0000000000883e9f rocksdb::BlockBasedTableBuilder::Add(rocksdb::Slice const&, rocksdb::Slice const&)
#12 0x0000000000a30b5e rocksdb::CompactionOutputs::AddToOutput(rocksdb::CompactionIterator const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&)> const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&, rocksdb::Status const&, rocksdb::Slice const&)> const&)
#13 0x0000000000a4ea6d rocksdb::SubcompactionState::AddToOutput(rocksdb::CompactionIterator const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&)> const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&, rocksdb::Status const&, rocksdb::Slice const&)> const&)
#14 0x0000000000a1f52c rocksdb::CompactionJob::ProcessKeyValueCompaction(rocksdb::SubcompactionState*)
#15 0x0000000000a243da rocksdb::CompactionJob::Run()
#16 0x000000000076a9ac rocksdb::DBImpl::BackgroundCompaction(bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#17 0x000000000076bf6f rocksdb::DBImpl::BackgroundCallCompaction(rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#18 0x000000000076c58e rocksdb::DBImpl::BGWorkCompaction(void*)
#19 0x000000000091e732 rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)
#20 0x000000000091e932 rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)
#21 0x00007f111272416f
#22 0x00007f111281eda4 start_thread
#23 0x00007f111248532d __clone
TID 1676:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x000000000084940d rocksdb::port::CondVar::Wait()
#2  0x0000000000a05f24 rocksdb::WriteThread::LinkOne(rocksdb::WriteThread::Writer*, std::atomic<rocksdb::WriteThread::Writer*>*)
#3  0x0000000000a063f6 rocksdb::WriteThread::JoinBatchGroup(rocksdb::WriteThread::Writer*)
#4  0x0000000000799cc7 rocksdb::DBImpl::PipelinedWriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, unsigned long*, unsigned long, bool, unsigned long*)
#5  0x000000000079c585 rocksdb::DBImpl::WriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, unsigned long*, unsigned long, bool, unsigned long*, unsigned long, rocksdb::PreReleaseCallback*, rocksdb::PostMemTableCallback*)
#6  0x000000000079dda0 rocksdb::DBImpl::Write(rocksdb::WriteOptions const&, rocksdb::WriteBatch*)
#7  0x00000000005a293c Engine::Storage::Write(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, bool)
#8  0x00000000005aa6a4 Engine::Storage::Delete(rocksdb::WriteOptions const&, rocksdb::ColumnFamilyHandle*, rocksdb::Slice const&, std::string const&)
#9  0x00000000005c5692 Engine::MetadataFilter::DeleteKeyIfExpired(rocksdb::Slice const&, Metadata*) const
#10 0x00000000005c5e6a Engine::MetadataFilter::Filter(int, rocksdb::Slice const&, rocksdb::Slice const&, std::string*, bool*) const
#11 0x0000000000a15356 rocksdb::CompactionIterator::InvokeFilterIfNeeded(bool*, rocksdb::Slice*)
#12 0x0000000000a17045 rocksdb::CompactionIterator::NextFromInput()
#13 0x0000000000a1a500 rocksdb::CompactionIterator::Next()
#14 0x0000000000a1f564 rocksdb::CompactionJob::ProcessKeyValueCompaction(rocksdb::SubcompactionState*)
#15 0x0000000000a243da rocksdb::CompactionJob::Run()
#16 0x000000000076a9ac rocksdb::DBImpl::BackgroundCompaction(bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#17 0x000000000076bf6f rocksdb::DBImpl::BackgroundCallCompaction(rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#18 0x000000000076c58e rocksdb::DBImpl::BGWorkCompaction(void*)
#19 0x000000000091e732 rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)
#20 0x000000000091e932 rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)
#21 0x00007f111272416f
#22 0x00007f111281eda4 start_thread
#23 0x00007f111248532d __clone
TID 1677:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x000000000084940d rocksdb::port::CondVar::Wait()
#2  0x000000000090e321 rocksdb::GenericRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)
#3  0x000000000090d72d rocksdb::RateLimiter::RequestToken(unsigned long, unsigned long, rocksdb::Env::IOPriority, rocksdb::Statistics*, rocksdb::RateLimiter::OpType)
#4  0x00000000007f08af rocksdb::WritableFileWriter::WriteBuffered(char const*, unsigned long, rocksdb::Env::IOPriority)
#5  0x00000000007f3659 rocksdb::WritableFileWriter::Flush(rocksdb::Env::IOPriority)
#6  0x00000000007f4a0d rocksdb::WritableFileWriter::Append(rocksdb::Slice const&, unsigned int, rocksdb::Env::IOPriority)
#7  0x000000000087f0af rocksdb::BlockBasedTableBuilder::WriteMaybeCompressedBlock(rocksdb::Slice const&, rocksdb::CompressionType, rocksdb::BlockHandle*, rocksdb::BlockType, rocksdb::Slice const*)
#8  0x000000000088127e rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*, rocksdb::BlockType)
#9  0x0000000000881706 rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::BlockBuilder*, rocksdb::BlockHandle*, rocksdb::BlockType)
#10 0x0000000000882ad2 rocksdb::BlockBasedTableBuilder::Flush()
#11 0x0000000000883e9f rocksdb::BlockBasedTableBuilder::Add(rocksdb::Slice const&, rocksdb::Slice const&)
#12 0x0000000000a30b5e rocksdb::CompactionOutputs::AddToOutput(rocksdb::CompactionIterator const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&)> const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&, rocksdb::Status const&, rocksdb::Slice const&)> const&)
#13 0x0000000000a4ea6d rocksdb::SubcompactionState::AddToOutput(rocksdb::CompactionIterator const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&)> const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&, rocksdb::Status const&, rocksdb::Slice const&)> const&)
#14 0x0000000000a1f52c rocksdb::CompactionJob::ProcessKeyValueCompaction(rocksdb::SubcompactionState*)
#15 0x0000000000a243da rocksdb::CompactionJob::Run()
#16 0x000000000076a9ac rocksdb::DBImpl::BackgroundCompaction(bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#17 0x000000000076bf6f rocksdb::DBImpl::BackgroundCallCompaction(rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#18 0x000000000076c58e rocksdb::DBImpl::BGWorkCompaction(void*)
#19 0x000000000091e732 rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)
#20 0x000000000091e932 rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)
#21 0x00007f111272416f
#22 0x00007f111281eda4 start_thread
#23 0x00007f111248532d __clone
TID 1678:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x000000000084940d rocksdb::port::CondVar::Wait()
#2  0x000000000090e321 rocksdb::GenericRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)
#3  0x000000000090d72d rocksdb::RateLimiter::RequestToken(unsigned long, unsigned long, rocksdb::Env::IOPriority, rocksdb::Statistics*, rocksdb::RateLimiter::OpType)
#4  0x00000000007f08af rocksdb::WritableFileWriter::WriteBuffered(char const*, unsigned long, rocksdb::Env::IOPriority)
#5  0x00000000007f3659 rocksdb::WritableFileWriter::Flush(rocksdb::Env::IOPriority)
#6  0x00000000007f4a0d rocksdb::WritableFileWriter::Append(rocksdb::Slice const&, unsigned int, rocksdb::Env::IOPriority)
#7  0x000000000087f0af rocksdb::BlockBasedTableBuilder::WriteMaybeCompressedBlock(rocksdb::Slice const&, rocksdb::CompressionType, rocksdb::BlockHandle*, rocksdb::BlockType, rocksdb::Slice const*)
#8  0x000000000088127e rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*, rocksdb::BlockType)
#9  0x0000000000881706 rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::BlockBuilder*, rocksdb::BlockHandle*, rocksdb::BlockType)
#10 0x0000000000882ad2 rocksdb::BlockBasedTableBuilder::Flush()
#11 0x0000000000883e9f rocksdb::BlockBasedTableBuilder::Add(rocksdb::Slice const&, rocksdb::Slice const&)
#12 0x0000000000a30b5e rocksdb::CompactionOutputs::AddToOutput(rocksdb::CompactionIterator const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&)> const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&, rocksdb::Status const&, rocksdb::Slice const&)> const&)
#13 0x0000000000a4ea6d rocksdb::SubcompactionState::AddToOutput(rocksdb::CompactionIterator const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&)> const&, std::function<rocksdb::Status (rocksdb::CompactionOutputs&, rocksdb::Status const&, rocksdb::Slice const&)> const&)
#14 0x0000000000a1f52c rocksdb::CompactionJob::ProcessKeyValueCompaction(rocksdb::SubcompactionState*)
#15 0x0000000000a243da rocksdb::CompactionJob::Run()
#16 0x000000000076a9ac rocksdb::DBImpl::BackgroundCompaction(bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#17 0x000000000076bf6f rocksdb::DBImpl::BackgroundCallCompaction(rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority)
#18 0x000000000076c58e rocksdb::DBImpl::BGWorkCompaction(void*)
#19 0x000000000091e732 rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)
#20 0x000000000091e932 rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)
#21 0x00007f111272416f
#22 0x00007f111281eda4 start_thread
#23 0x00007f111248532d __clone
TID 1679:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x000000000084940d rocksdb::port::CondVar::Wait()
#2  0x000000000090e321 rocksdb::GenericRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)
#3  0x000000000090d72d rocksdb::RateLimiter::RequestToken(unsigned long, unsigned long, rocksdb::Env::IOPriority, rocksdb::Statistics*, rocksdb::RateLimiter::OpType)
#4  0x00000000007f08af rocksdb::WritableFileWriter::WriteBuffered(char const*, unsigned long, rocksdb::Env::IOPriority)
#5  0x00000000007f3659 rocksdb::WritableFileWriter::Flush(rocksdb::Env::IOPriority)
#6  0x00000000007f4a0d rocksdb::WritableFileWriter::Append(rocksdb::Slice const&, unsigned int, rocksdb::Env::IOPriority)
#7  0x000000000087f0af rocksdb::BlockBasedTableBuilder::WriteMaybeCompressedBlock(rocksdb::Slice const&, rocksdb::CompressionType, rocksdb::BlockHandle*, rocksdb::BlockType, rocksdb::Slice const*)
#8  0x000000000088127e rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*, rocksdb::BlockType)
#9  0x0000000000881706 rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::BlockBuilder*, rocksdb::BlockHandle*, rocksdb::BlockType)
#10 0x0000000000882ad2 rocksdb::BlockBasedTableBuilder::Flush()
#11 0x0000000000883e9f rocksdb::BlockBasedTableBuilder::Add(rocksdb::Slice const&, rocksdb::Slice const&)
#12 0x0000000000980895 rocksdb::BuildTable(std::string const&, rocksdb::VersionSet*, rocksdb::ImmutableDBOptions const&, rocksdb::TableBuilderOptions const&, rocksdb::FileOptions const&, rocksdb::TableCache*, rocksdb::InternalIteratorBase<rocksdb::Slice>*, std::vector<std::unique_ptr<rocksdb::FragmentedRangeTombstoneIterator, std::default_delete<rocksdb::FragmentedRangeTombstoneIterator> >, std::allocator<std::unique_ptr<rocksdb::FragmentedRangeTombstoneIterator, std::default_delete<rocksdb::FragmentedRangeTombstoneIterator> > > >, rocksdb::FileMetaData*, std::vector<rocksdb::BlobFileAddition, std::allocator<rocksdb::BlobFileAddition> >*, std::vector<unsigned long, std::allocator<unsigned long> >, unsigned long, unsigned long, rocksdb::SnapshotChecker*, bool, rocksdb::InternalStats*, rocksdb::IOStatus*, std::shared_ptr<rocksdb::IOTracer> const&, rocksdb::BlobFileCreationReason, rocksdb::SeqnoToTimeMapping const&, rocksdb::EventLogger*, int, rocksdb::Env::IOPriority, rocksdb::TableProperties*, rocksdb::Env::WriteLifeTimeHint, std::string const*, rocksdb::BlobFileCompletionCallback*, rocksdb::Version*, unsigned long*, unsigned long*, unsigned long*)
#13 0x00000000009ba483 rocksdb::FlushJob::WriteLevel0Table()
#14 0x00000000009bc5ef rocksdb::FlushJob::Run(rocksdb::LogsWithPrepTracker*, rocksdb::FileMetaData*, bool*)
#15 0x000000000075ea67 rocksdb::DBImpl::FlushMemTableToOutputFile(rocksdb::ColumnFamilyData*, rocksdb::MutableCFOptions const&, bool*, rocksdb::JobContext*, rocksdb::FlushReason, rocksdb::SuperVersionContext*, std::vector<unsigned long, std::allocator<unsigned long> >&, unsigned long, rocksdb::SnapshotChecker*, rocksdb::LogBuffer*, rocksdb::Env::Priority)
#16 0x0000000000761e9d rocksdb::DBImpl::FlushMemTablesToOutputFiles(rocksdb::autovector<rocksdb::DBImpl::BGFlushArg, 8ul> const&, bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::Env::Priority)
#17 0x0000000000762984 rocksdb::DBImpl::BackgroundFlush(bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::FlushReason*, rocksdb::Env::Priority)
#18 0x00000000007680e8 rocksdb::DBImpl::BackgroundCallFlush(rocksdb::Env::Priority)
#19 0x000000000091e732 rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)
#20 0x000000000091e932 rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)
#21 0x00007f111272416f
#22 0x00007f111281eda4 start_thread
#23 0x00007f111248532d __clone
TID 1682:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x000000000084940d rocksdb::port::CondVar::Wait()
#2  0x000000000090e321 rocksdb::GenericRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)
#3  0x000000000090d72d rocksdb::RateLimiter::RequestToken(unsigned long, unsigned long, rocksdb::Env::IOPriority, rocksdb::Statistics*, rocksdb::RateLimiter::OpType)
#4  0x00000000007f08af rocksdb::WritableFileWriter::WriteBuffered(char const*, unsigned long, rocksdb::Env::IOPriority)
#5  0x00000000007f3659 rocksdb::WritableFileWriter::Flush(rocksdb::Env::IOPriority)
#6  0x00000000007f4a0d rocksdb::WritableFileWriter::Append(rocksdb::Slice const&, unsigned int, rocksdb::Env::IOPriority)
#7  0x000000000087f0af rocksdb::BlockBasedTableBuilder::WriteMaybeCompressedBlock(rocksdb::Slice const&, rocksdb::CompressionType, rocksdb::BlockHandle*, rocksdb::BlockType, rocksdb::Slice const*)
#8  0x000000000088127e rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*, rocksdb::BlockType)
#9  0x0000000000881706 rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::BlockBuilder*, rocksdb::BlockHandle*, rocksdb::BlockType)
#10 0x0000000000882ad2 rocksdb::BlockBasedTableBuilder::Flush()
#11 0x0000000000883e9f rocksdb::BlockBasedTableBuilder::Add(rocksdb::Slice const&, rocksdb::Slice const&)
#12 0x0000000000980895 rocksdb::BuildTable(std::string const&, rocksdb::VersionSet*, rocksdb::ImmutableDBOptions const&, rocksdb::TableBuilderOptions const&, rocksdb::FileOptions const&, rocksdb::TableCache*, rocksdb::InternalIteratorBase<rocksdb::Slice>*, std::vector<std::unique_ptr<rocksdb::FragmentedRangeTombstoneIterator, std::default_delete<rocksdb::FragmentedRangeTombstoneIterator> >, std::allocator<std::unique_ptr<rocksdb::FragmentedRangeTombstoneIterator, std::default_delete<rocksdb::FragmentedRangeTombstoneIterator> > > >, rocksdb::FileMetaData*, std::vector<rocksdb::BlobFileAddition, std::allocator<rocksdb::BlobFileAddition> >*, std::vector<unsigned long, std::allocator<unsigned long> >, unsigned long, unsigned long, rocksdb::SnapshotChecker*, bool, rocksdb::InternalStats*, rocksdb::IOStatus*, std::shared_ptr<rocksdb::IOTracer> const&, rocksdb::BlobFileCreationReason, rocksdb::SeqnoToTimeMapping const&, rocksdb::EventLogger*, int, rocksdb::Env::IOPriority, rocksdb::TableProperties*, rocksdb::Env::WriteLifeTimeHint, std::string const*, rocksdb::BlobFileCompletionCallback*, rocksdb::Version*, unsigned long*, unsigned long*, unsigned long*)
#13 0x00000000009ba483 rocksdb::FlushJob::WriteLevel0Table()
#14 0x00000000009bc5ef rocksdb::FlushJob::Run(rocksdb::LogsWithPrepTracker*, rocksdb::FileMetaData*, bool*)
#15 0x000000000075ea67 rocksdb::DBImpl::FlushMemTableToOutputFile(rocksdb::ColumnFamilyData*, rocksdb::MutableCFOptions const&, bool*, rocksdb::JobContext*, rocksdb::FlushReason, rocksdb::SuperVersionContext*, std::vector<unsigned long, std::allocator<unsigned long> >&, unsigned long, rocksdb::SnapshotChecker*, rocksdb::LogBuffer*, rocksdb::Env::Priority)
#16 0x0000000000761e9d rocksdb::DBImpl::FlushMemTablesToOutputFiles(rocksdb::autovector<rocksdb::DBImpl::BGFlushArg, 8ul> const&, bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::Env::Priority)
#17 0x0000000000762984 rocksdb::DBImpl::BackgroundFlush(bool*, rocksdb::JobContext*, rocksdb::LogBuffer*, rocksdb::FlushReason*, rocksdb::Env::Priority)
#18 0x00000000007680e8 rocksdb::DBImpl::BackgroundCallFlush(rocksdb::Env::Priority)
#19 0x000000000091e732 rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)
#20 0x000000000091e932 rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)
#21 0x00007f111272416f
#22 0x00007f111281eda4 start_thread
#23 0x00007f111248532d __clone
TID 1776:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x00007f111271e9fc std::condition_variable::wait(std::unique_lock<std::mutex>&)
#2  0x0000000000a05aeb rocksdb::WriteThread::BlockingAwaitState(rocksdb::WriteThread::Writer*, unsigned char)
#3  0x0000000000a05c27 rocksdb::WriteThread::AwaitState(rocksdb::WriteThread::Writer*, unsigned char, rocksdb::WriteThread::AdaptationContext*)
#4  0x0000000000799cc7 rocksdb::DBImpl::PipelinedWriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, unsigned long*, unsigned long, bool, unsigned long*)
#5  0x000000000079c585 rocksdb::DBImpl::WriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, unsigned long*, unsigned long, bool, unsigned long*, unsigned long, rocksdb::PreReleaseCallback*, rocksdb::PostMemTableCallback*)
#6  0x000000000079dda0 rocksdb::DBImpl::Write(rocksdb::WriteOptions const&, rocksdb::WriteBatch*)
#7  0x00000000005a293c Engine::Storage::Write(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, bool)
#8  0x0000000000521985 Redis::Hash::IncrBy(rocksdb::Slice const&, rocksdb::Slice const&, long, long*)
#9  0x00000000004e9552 Redis::CommandHIncrBy::Execute(Server*, Redis::Connection*, std::string*)
#10 0x000000000050dc3f Redis::Connection::ExecuteCommands(std::vector<std::vector<std::string, std::allocator<std::string> >, std::allocator<std::vector<std::string, std::allocator<std::string> > > > const&)
#11 0x000000000050ec8e Redis::Connection::OnRead(bufferevent*, void*)
#12 0x0000000000b91b0e bufferevent_run_deferred_callbacks_unlocked
#13 0x0000000000b9a12a event_process_active_single_queue
#14 0x0000000000b9abd7 event_base_loop
#15 0x00000000005b836a Worker::Run(std::thread::id)
#16 0x00007f111272416f
#17 0x00007f111281eda4 start_thread
#18 0x00007f111248532d __clone
...
TID 1787:
#0  0x00007f111282439f pthread_cond_wait@@GLIBC_2.3.2
#1  0x00007f111271e9fc std::condition_variable::wait(std::unique_lock<std::mutex>&)
#2  0x0000000000a05aeb rocksdb::WriteThread::BlockingAwaitState(rocksdb::WriteThread::Writer*, unsigned char)
#3  0x0000000000a05c27 rocksdb::WriteThread::AwaitState(rocksdb::WriteThread::Writer*, unsigned char, rocksdb::WriteThread::AdaptationContext*)
#4  0x0000000000799cc7 rocksdb::DBImpl::PipelinedWriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, unsigned long*, unsigned long, bool, unsigned long*)
#5  0x000000000079c585 rocksdb::DBImpl::WriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, unsigned long*, unsigned long, bool, unsigned long*, unsigned long, rocksdb::PreReleaseCallback*, rocksdb::PostMemTableCallback*)
#6  0x000000000079dda0 rocksdb::DBImpl::Write(rocksdb::WriteOptions const&, rocksdb::WriteBatch*)
#7  0x00000000005a293c Engine::Storage::Write(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, bool)
#8  0x00000000005635f3 Redis::ZSet::Add(rocksdb::Slice const&, unsigned char, std::vector<MemberScore, std::allocator<MemberScore> >*, int*, bool)
#9  0x00000000004ee56e Redis::CommandZAdd::Execute(Server*, Redis::Connection*, std::string*)
#10 0x000000000050dc3f Redis::Connection::ExecuteCommands(std::vector<std::vector<std::string, std::allocator<std::string> >, std::allocator<std::vector<std::string, std::allocator<std::string> > > > const&)
#11 0x000000000050ec8e Redis::Connection::OnRead(bufferevent*, void*)
#12 0x0000000000b91b0e bufferevent_run_deferred_callbacks_unlocked
#13 0x0000000000b9a12a event_process_active_single_queue
#14 0x0000000000b9abd7 event_base_loop
#15 0x00000000005b836a Worker::Run(std::thread::id)
#16 0x00007f111272416f
#17 0x00007f111281eda4 start_thread
#18 0x00007f111248532d __clone
```

### PR DESCRIPTION
**Pull request by ajkr**
Fixes #11742

Even after performing duty (1) ("Waiting for the next refill time"), it is possible the remaining threads are all in `Wait()`. Waking up at least one thread is enough to ensure progress continues, even if no new requests arrive.

The repro unit test (https://github.com/facebook/rocksdb/commit/bb54245e6) is not included as it depends on an unlanded PR (#11753)

### CODE REVIEW

**Review by jaykorean — APPROVED** [review]
Thanks for the fix!

File: util/rate_limiter.cc [hunk 1]
```diff
@@ -179,16 +179,16 @@ void GenericRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
...
-            queue.front()->cv.Signal();
-            break;
-          }
+    }
+    if (r.request_bytes == 0) {
+      // If there is any remaining requests, make sure there exists at least
+      // one candidate is awake for future duties by signaling a front request
+      // of a queue.
+      for (int i = Env::IO_TOTAL - 1; i >= Env::IO_LOW; --i) {
+        std::deque<Req*> queue = queue_[i];
```
Comment (ronag) [inline util/rate_limiter.cc:188]:
```suggestion
        auto& queue = queue_[i];
```

Or you will be copying it?

Comment (ajkr) [reply util/rate_limiter.cc]: Hm yeah, I didn't notice that - thanks

**Comment by ajkr** [comment]
Thanks for the review!

**Comment by hx235** [comment]
So the thread 2 did the granting job, which a previous thread was supposed to do if waking up exactly from TimedWait(), and made the only thread (the prev thread) that can wake itself (thread 2) up exit. While we relaxed who can grant bytes after https://github.com/facebook/rocksdb/pull/8602, we didn't relax who can wake up remaining requests - it was only the thread granting bytes can wake up remaining requests. Interesting findings!

**Pull request closed without merging**